MerlinAU.sh

```diff
@@ -5163,14 +5755,16 @@ Please manually update to version $MinSupportedFirmwareVers or higher to use thi
     ##------------------------------------------##
     ## Modified by ExtremeFiretop [2024-May-25] ##
     ##------------------------------------------##
-    _ChangelogVerificationCheck_ "interactive"
-    retCode="$?"
+    if ! [[ "$isGNUtonFW" && "$offlineUpdateTrigger" ]]; then
```

Comment:
Line **5758**: using double square brackets `"[[ ... ]]`" to evaluate conditional expressions is considered non-standard for the Ash shell and is not consistent with the rest of the code.

Use the following syntax instead:

```
if ! { "$isGNUtonFW" && "$offlineUpdateTrigger" ; }
```

Reply:
I'm off to bed, but will correct anything mentioned in the morning!

Thanks again Martinski!

Reply:
Have a goodnight, we'll chat more tomorrow evening, I'll try to get that list of functions together

Reply:
> Have a goodnight, we'll chat more tomorrow evening, I'll try to get that list of functions together
>
Yep, good night, bud! I'm going to bed in a few minutes as well. We'll continue tomorrow evening.

Reply:
Corrected!

Reply:
> Corrected!
>
Apologies for the late replies. My brother & his wife stopped by this evening so we all went out to eat and then watched a movie.

Anyway, see my post-original message WRT line **5758** above:
>
>
> P.S.
> After a 2nd look, now I'm curious.
> You do not want to run the "Changelog Verification Check" during "offline" updates for Gnuton F/W.
> Is that because it's no longer required to have a changelog for Gnuton during offline updates?
> 
Just wondering what the reason is. I'm not really familiar with Gnuton so I don't have enough context to fully understand here.

Reply:
@Martinski4GitHub 

No problem on the delay, as for the Gnuton specific question.
Have you had a look at the Gnuton branch, 🥺 Hahaha
I think if you plan on doing a review, will help just understand the changes instead of assuming why they are done.

For Gnuton, the files come all separately and not in a zip.
I'm simply not going to make a Gnuton user drop 3 different separate files into the offline directory to do an offline flash. A changelog file is not required to flash, and neither is a md5 file.

If they want to flash firmware offline, all we need is the firmware, that's all me and you need to do an offline flash as well when we do it through the webUI, I'm not going to be invoking extra requirements that aren't using required.

In Merlin we can do it because it's all packaged together anyways, but with Gnuton absolutely not instead you work with the assumption all they have downloaded is the rew image file. (Which is correct 99.9% of the time)

We know this feature is excluding some safeties and it's even documented when you start the script.

Reply:
@Martinski4GitHub 

With all the above said. Now I'm second guessing if there's value in having them drop the md5 file though... Any thoughts? The changelogs on Gnuton are pretty useless. It only contains the data of that specific release. So usually like 3 or 4 lines of info lol.

But maybe forcing them to drop the .md5 file has value? Thoughts?
Your making me second guess myself now...

Reply:
Would the .md5 being dropped be a separate screen than the zip being dropped? Or should we make them drop it all in the one screen?

Reply:
It doesn't address the fact that we would likely be asking them to go back to Gnuton to download the .md5 file because they likely didn't do that to start though... Idk how I feel about it.

Reply:
Upon further reflection, I think the md5 file might have value. But I still think the changelog check can be skipped for Gnuton.

Check the Gnuton branchs changelog notes, they usually contains little and they are posted with the firmware raw file so I doubt anyone downloads it. But we could request a user download the md5 file?

Reply:
> @Martinski4GitHub
> 
> _EDITED MULTIPLE TIMES TO MAKE THE TEXT AND TONE MORE CLEAR_
> 
> No problem on the delay, as for the Gnuton specific question, sorry for my delay! I didn't even see it.
> 
> Have you had a look at the Gnuton branch, 🥺 Hahaha I think if you plan on doing a review, will help just understand the changes😉 But for Gnuton, the files come all separately and not in a zip.
> 
>
OK, now I understand. That's the piece of info I was missing all along: the 3 files that are zipped together in RMerlin's F/W are all available **separately** to download with Gnuton F/W (I wonder why he doesn't zip them up).


>
> I'm simply not going to ask a Gnuton user drop 3 different separate files into the offline directory to do an offline flash. A changelog file is not required to flash, and neither is a md5 file. We know this feature is excluding some safeties and it's documented when you start the script.
> 
>
Yeah, the underlying assumption is that the new feature is for "advanced users" who presumably know & understand the risks involved when doing the offline update, so if Gnuton does not offer a ZIP file then the hash signature verification and the changelog check are pretty much off the table. Also, the assumption is that this feature is mostly for people who would like to install Alpha & Beta releases, which also have their own risks.

Reply:
We could make it an "optional" upload?

Aka. if they upload it. We check it. Otherwise skip it?

Reply:
> @Martinski4GitHub
> 
> With all the above said. Now I'm second guessing if there's value in having them drop the md5 file though... Any thoughts? The changelogs on Gnuton are pretty useless. It only contains the data of that specific release. So usually like 3 or 4 lines of info lol.
> 
> But maybe forcing them to drop the .md5 file has value? Thoughts? Your making me second guess myself now...
>
I suppose it could be added as an option, but since this is an offline update anyway where the source of the image file is already considered "unvetted" (as far as the script is concerned), asking for a separate file for the hash signature seems like a moot point given that it's of "unknown origin."

Reply:
> Upon further reflection, I think the md5 file might have value. But I still think the changelog check can be skipped for Gnuton.
> 
> Check the Gnuton branchs changelog notes, they usually contains little and they are posted with the firmware raw file so I doubt anyone downloads it. 
>
Yeah, the Gnuton's change log has little to no valuable information. I thought he would just add to the one coming from RMerlin's for the equivalent F/W version, but I guess some of it might not apply given the different router models that Gnuton supports.

>
> But we could request a user download the md5 file? Opinions?
>
See my previous comment about this. IMO, given that the files (image & hash signature) come from an "unknown/unvetted" source (from the perspective of the script), I see very little value in asking the user to have an MD5 signature to check.

My 2 cents.

Reply:
> We could make it an "optional" upload?
> 
> Aka. if they upload it. We check it. Otherwise skip it?
>
>
So you mean the process would not explicitly ask the user to download the MD5 signature file, but if the script finds it along with the image file, we do the hash verification?
If so, that's certainly a way to do it without forcing the user to download it every time.

Reply:
I think because all the code exists to check the md5 file already. It wouldn't be hard to add a conditional if they upload an md5 file to check it. Otherwise skip it. Make it optional but not required for an offline flash.

Reply:
> > We could make it an "optional" upload?
> > Aka. if they upload it. We check it. Otherwise skip it?
> 
> So you mean the process would not explicitly ask the user to download the MD5 signature file, but if the script finds it along with the image file, we do the hash verification? If so, that's certainly a way to do it without forcing the user to download it every time.

Exactly! Kinda like a hybrid of both approaches? Handles my concern of forcing users to go get a file they probably didn't get.

But if they are smart enough to think about the md5 check, we check it for them. It's one less risk to worry about in that regard right?

Reply:
> > > We could make it an "optional" upload?
> > > Aka. if they upload it. We check it. Otherwise skip it?
> > 
> > 
> > So you mean the process would not explicitly ask the user to download the MD5 signature file, but if the script finds it along with the image file, we do the hash verification? If so, that's certainly a way to do it without forcing the user to download it every time.
> 
> Exactly! Kinda like a hybrid of both approaches? Handles my concern of forcing users to go get a file they probably didn't get.
> 
> But if they are smart enough to think about the md5 check, we check it for them. It's one less risk to worry about in that regard right?
>
>
Yeah, that would work for both use cases: users who don't care to download it to be checked, and those users who do.

Reply:
> > > > We could make it an "optional" upload?
> > > > Aka. if they upload it. We check it. Otherwise skip it?
> > > 
> > > 
> > > So you mean the process would not explicitly ask the user to download the MD5 signature file, but if the script finds it along with the image file, we do the hash verification? If so, that's certainly a way to do it without forcing the user to download it every time.
> > 
> > 
> > Exactly! Kinda like a hybrid of both approaches? Handles my concern of forcing users to go get a file they probably didn't get.
> > But if they are smart enough to think about the md5 check, we check it for them. It's one less risk to worry about in that regard right?
> 
> Yeah, that would work for both use cases: users who don't care to download it to be checked, and those users who do.

Done in commit: https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/pull/186/commits/a0af4e402e174083eca3eaafe0fa4c7f28fed167